torch/optim/functional.py

```diff
@@ -0,0 +1,46 @@
+r"""Functional interface"""
+from torch import Tensor
+from typing import List
+
+# TODO: use foreach API in optim.functional to do all the computation
+
+def adagrad(params: List[Tensor],
```

Comment:
we should have a test that gives some guarantees we have not diverged from the original implementation.

Reply:
currently I am relying on the original optimizer tests to test the coverage. https://github.com/pytorch/pytorch/blob/24eea364f761bc841cb3f3890d1ce0200571856e/test/test_optim.py#L384 Are there any new cases that you think we need to add to give the guarantees?

Reply:
Since we are replacing the current implementation, the C++ tests confirm that the the python and C++ give the same answers, see [comment](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986). This gives the signal we need here.

cc @anjali411